lib/FrontendTool/LoadedModuleTrace.cpp

```diff
@@ -810,3 +813,69 @@ bool swift::emitLoadedModuleTraceIfNeeded(ModuleDecl *mainModule,
   }
   return false;
 }
+
+
+bool swift::emitObjCMessageSendTraceIfNeeded(ModuleDecl *mainModule,
+                                             const FrontendOptions &opts) {
+  ASTContext &ctxt = mainModule->getASTContext();
+  assert(!ctxt.hadError() &&
+         "We should've already exited earlier if there was an error.");
+  class ObjcMethodRefereceCollector: public SourceEntityWalker {
+    std::set<const clang::ObjCMethodDecl*> results;
+    bool visitDeclReference(ValueDecl *D, CharSourceRange Range,
+                            TypeDecl *CtorTyRef, ExtensionDecl *ExtTyRef,
+                            Type T, ReferenceMetaData Data) override {
+      if (!Range.isValid())
+        return true;
+      if (auto *clangD = dyn_cast_or_null<clang::ObjCMethodDecl>(D->getClangDecl())) {
+        results.insert(clangD);
+      }
+      return true;
+    }
+  public:
+    void dump(llvm::raw_ostream &OS) {
+      OS << "[\n";
+      for (const clang::ObjCMethodDecl* clangD: results) {
+        auto &SM = clangD->getASTContext().getSourceManager();
+        clang::SourceLocation Loc = clangD->getLocation();
+        if (!Loc.isValid()) {
+          continue;
+        }
+        OS << "\t{\n";
+        OS << "\t\t\"method_name\": \"" << clangD->getNameAsString() << "\",\n";
+        OS << "\t\t\"location\": \"" << Loc.printToString(SM) << "\"\n";
+        OS << "\t}";
+        OS << ",\n";
+      }
+      OS << "{} ]\n";
+    }
+  };
+  opts.InputsAndOutputs.forEachInput([&](const InputFile &input) {
+    auto loadedModuleTracePath = input.getLoadedModuleTracePath();
+    if (loadedModuleTracePath.empty())
+      return false;
+    llvm::SmallString<128> tracePath {loadedModuleTracePath};
+    llvm::sys::path::remove_filename(tracePath);
+    llvm::sys::path::append(tracePath, "SWIFT_OBJC_MESSAGE_TRACE");
```

Comment:
Caps is a little weird but module trace does the same so 🤷. Maybe `.SWIFT_OBJC_MESSAGE_TRACE` then though?